Distinguish client and internal errors in action responses

Annaseli opened the pull request:
## Change Description

Phase one of addressing the issue #9147

### Issue
When a user ran pipelines, they encountered a 412 (http.StatusPreconditionFailed) status code. This occurred because cosmosdb's throughput was exceeded, causing graveler to return a kv.ErrSlowDown error. That error was wrapped in a graveler.HookAbortError, which caused a 412 Precondition Failed (http.StatusPreconditionFailed) status code to be returned instead of a more specific lakeFS internal error status code.

### Fix Description
Currently, when an error comes from actions, we return a 412 (http.StatusPreconditionFailed) status code, even if the error was caused by an internal lakeFS issue. As a result, users don’t receive an accurate status code for internal errors.
The goal of this PR is to distinguish between status codes returned from actions: separating those caused by internal lakeFS errors from those caused by client-side action errors.

### Fix Implementation
Refactored error handling in the controller to handle `actions.ErrActionFailed` as a distinct case that returns an HTTP 412 status code. This new error type is used to wrap client-side action errors, allowing the controller to differentiate them from internal lakeFS errors resulting from failed actions.


### Testing Details

Tested local lakeFS with local KV:
1. Added an action file to lakeFS named check_create_branch.yml.
2. Created a new branch and verified that it was created successfully and the hook executed as expected.
3. Modified the Get function in local/store.go to return a kv.ErrSlowDown error when the composed key (from partitionKey and key) contains the string "_lakefs_actions/check_create_branch.yml": 
![image](https://github.com/user-attachments/assets/06bac967-3e60-4746-8d92-206676aeca83)

4. Created a new branch via the UI.
5. As expected, the server returned a 503 status code, instead of allowing branch creation and hook execution:

lakeFS logs:

```
found the key!
time="2025-06-08T20:33:58+03:00" level=info msg="HTTP call ended" func=net/http.HandlerFunc.ServeHTTP file="Users/annaseliverstov/go/go1.23.0/src/net/http/server.go:2220" client=lakefs-webui/dev host="localhost:8000" log_audit=true method=POST operation_id=CreateBranch path=/api/v1/repositories/test-actions-sourc-412-issue/branches request_id=880038b7-cf33-401e-a280-028c7916761d sent_bytes=65 service_name=rest_api source_ip="::1:50574" status_code=503 took=31087542 took_str=31.087542ms user=admin

```

In the UI:
![image](https://github.com/user-attachments/assets/229ad764-1bb5-4f49-b095-60a6b8907f24)

itaiad200 commented:
Adding @N-o-Z / @nopcoder / @arielshaqed , if any one of you could review too that'd be great

itaiad200 commented on pkg/api/controller.go:
```diff
@@ -3016,7 +3022,7 @@ func (c *Controller) handleAPIErrorCallback(ctx context.Context, w http.Response
 	case errors.Is(err, kv.ErrSlowDown):
 		log.Debug("KV Throttling")
-		cb(w, r, http.StatusServiceUnavailable, "Throughput exceeded. Slow down and retry")
+		cb(w, r, http.StatusServiceUnavailable, "Key value throughput exceeded. Slow down and retry")
```
Though the edited error message is better, even changing the body of an error response is considered a breaking change to our 1.0 guarantees, so we need to revert this one.

itaiad200 reviewed (changes requested):
Thank you for the quick fix! Blocking for the minor, yet important, breaking change

I think this works. I requested one more opinion since the solution might overfit out specific issue. I wonder if there's a better way to generalize hooks that might fail due to errors that are unrelated to the hooks themselves (i.e. unrelated to run failures or even parsing errors)

nopcoder reviewed (commented):
we can capture the reason by reporting the original error after checking the type, and still abort with precondition error. the idea was to report that the commit failed because of an action failed.

this fix still reports an error, the issue was the error message we report?

N-o-Z commented:
Why did we frame this task as return 503? We need to return a terrible http code (e.g. 429)

arielshaqed reviewed (commented):
Thanks!  LGTM.  I have no blocking comments; I am not approving because it seems the return code is still somewhat open.

Annaseli commented on pkg/api/controller.go:
```diff
@@ -2937,6 +2937,12 @@ func (c *Controller) handleAPIErrorCallback(ctx context.Context, w http.Response
 	var hookAbortErr *graveler.HookAbortError
 	if errors.As(err, &hookAbortErr) {
+		if errors.Is(hookAbortErr.Unwrap(), kv.ErrSlowDown) {
+			log.Debug("KV Throttling")
```
got it. so changed the "log.Debug" to:

`
log.Warn("Hooks aborted by KV throttling")`

Indeed I can see the "run id" in the error itself.

Annaseli commented:
@itaiad200 @nopcoder @arielshaqed @N-o-Z 

### Changes made:

1. Added a new error wrapper called "HookClientError" to distinguish between client-related action errors from internal lakeFS errors.
2. Wrapped client-related action errors in the "Run" function in these files: lua.go, webhook.go, airflow.go
3. Updated the controller to check whether a graveler.HookAbortError wraps an actions.HookClientError. If so, it returns a 412 status. Otherwise, it proceeds to the existing switch block for internal lakeFS errors.

### Testing details:

Tested manually locally using lakeFS with a local KV store:
1. Uploaded an action file called check_create_branch.yml to run on pre-create-branch.
[check_create_branch.txt](https://github.com/user-attachments/files/20661253/check_create_branch.txt)

2. Simulated internal errors by modifying the "Get" function in local/store.go.
* Attempted to create a new branch and verified that it failed as expected.
* Confirmed the correct status code was returned for internal errors, as defined in the handleAPIErrorCallback function in controller.go.
* Reverted the changes made to Get after testing.

3. Injected internal errors directly in the Run function in lua.go.
* Verified that branch creation failed and the appropriate internal status code was returned.
* Reverted these changes after testing.

4. Uploaded the following action files (one by one) to trigger client-side errors and validated that the system returned a 412 status code as expected:
* calls_error.yml: Lua script calls error() explicitly.
[calls_error.txt](https://github.com/user-attachments/files/20661285/calls_error.txt)
* synt_err.yml: Lua script contains a syntax error.
[synt_err.txt](https://github.com/user-attachments/files/20661303/synt_err.txt)
* api_call.yml:
   * Requires the lakefs module.
   * Calls lakefs.list_objects() on my local main branch.
   * Attempts to get_object() on a non-existent object.
   * Returns a client error if get_object() fails.
[api_call.txt](https://github.com/user-attachments/files/20661329/api_call.txt)
* All client-related error scenarios correctly returned 412 status with appropriate log output.

* used these action files as yml and not txt as uploaded here.

nopcoder commented on pkg/actions/errors.go:
can be simplified by just having an ErrActionFailed error that will wrap the errors we accept from invoking the action itself.

nopcoder commented on pkg/api/controller.go:
```diff
@@ -2937,9 +2937,14 @@ func (c *Controller) handleAPIErrorCallback(ctx context.Context, w http.Response
-		log.WithField("run_id", hookAbortErr.RunID).Warn("aborted by hooks")
-		cb(w, r, http.StatusPreconditionFailed, hookAbortErr.Unwrap())
-		return true
+		var clientErr *actions.HookClientError
```
- if we use fmt to wrap the errors from the execution, we don't need this code.
also we like to report the full error not just the underlying one as we miss information about the action itself.

nopcoder reviewed (changes requested):
suggest to simplify the code by using an error to wrap the specific cases.

Suggestion, I know that I suggested to minimize the scope of the client error to the actual call we perform in the action execution, but maybe it will be simpler to wrap the error returning from Run itself as most of the code there is related to the actual action execution without lakeFS related operations.

itaiad200 reviewed (approved):
LGTM - thanks for the thorough manual tests! We should do something about the lack of ability to simulate errors during tests - but not as part of this PR.

+1 on @nopcoder comments, but none of my own

itaiad200 commented:
> > > > Why did we frame this task as return 503? We need to return a retryable http code (e.g. 429)
> > > 
> > > 
> > > Isn't 503 retryable? Notice that we're keeping the existing behavior:
> > > https://github.com/treeverse/lakeFS/blob/62e350d8c55da40923f975dcd7e3269c6bba953d/pkg/api/controller.go#L3017
> > 
> > 
> > I noticed that we already handle `kv.ErrSlowDown` as `ServiceUnavaliable`. However, in our swagger we define `420` throughout our API as 'too many requests'. We should align the swagger with the controller code.We should either use 420 or 503 or define explicitly when each should be returned
> 
> In fact, we have a broader inconsistency. In status.go (from the http package), several status codes used in our codebase **are not declared at all** in the Swagger spec. Example: 503 StatusServiceUnavailablee, 300 StatusMultipleChoices, and others. On the other hand, **these are the only non-2xx response status codes defined in the Swagger file:**
> 
> 1. 302: Successfully got token
> 2. 304: Content not modified
> 3. 400: BadRequest / ValidationError
> 4. 401: Unauthorized
> 5. 403: Forbidden
> 6. 404: NotFound / NotFoundOrNoACL
> 7. 409: Conflict
> 8. 410: Object expired
> 9. 412: Precondition failed
> 10. 416: Requested range not satisfiable
> 11. 420: Too many requests
> 12. 500: Server error (default)
> 13. 501: Not implemented
> 
> Also some of these are defined using **components.responses** (e.g., 404 NotFoundOrNoACL) in the Swagger spec, while others like 420 are defined inline without a shared response schema. for example:
> 
> ```
> 404:
>           $ref: "#/components/responses/NotFoundOrNoACL"
> NotFoundOrNoACL:
>       description: Group not found, or group found but has no ACL
>       content:
>         application/json:
>           schema:
>             $ref: "#/components/schemas/ErrorNoACL"
> ```
> 
> ```
> 420:
>           description: too many requests
> ```
> 
> There are also **duplicate status codes** with different descriptions (400 and 404).
> 
> **Additional inconsistencies:**
> 
> 1. Same as with 503 - we return 423 StatusLocked, but it’s not mentioned in Swagger:
> 
> ```
> case errors.Is(err, graveler.ErrTooManyTries):
> 		log.Debug("Retried too many times")
> 		cb(w, r, http.StatusLocked, "Too many attempts, try again later")
> ```
> 
> 2. We return 410 StatusGone with the message “No data”, but in Swagger it’s described as “object expired”:
> 
> ```
> case errors.Is(err, block.ErrDataNotFound):
> 		log.Debug("No data")
> 		cb(w, r, http.StatusGone, "No data")
> ```
> 
> **Question:** if we change the Swagger file or modify the status codes or their messages in the code, would that be considered a breaking change?
> 
> @itaiad200 @N-o-Z How would you like me to continue? should I modify something in this PR? should I open a new issue? If to open an issue, should it address changing the entire Swagger file's status codes with our actual API behavior? or to change specific endpoints's status codes or something else?

Very important find!
It shouldn't block this PR. Please track it in an issue, we can decide later on how to fix this.

nopcoder commented on pkg/api/controller.go:
```diff
@@ -2936,9 +2936,9 @@ func (c *Controller) handleAPIErrorCallback(ctx context.Context, w http.Response
-	if errors.As(err, &hookAbortErr) {
-		log.WithField("run_id", hookAbortErr.RunID).Warn("aborted by hooks")
-		cb(w, r, http.StatusPreconditionFailed, hookAbortErr.Unwrap())
+	if errors.As(err, &hookAbortErr) && errors.Is(hookAbortErr, actions.ErrActionFailed) {
```
note that we have a specific error for action failed, we can add a case in the switch to return predicate failed for this error and remove this code.

nopcoder commented:
if I remember the 42x was added when we supported the s3 throttle in the gateway and we said that we will need to do the same for the API. Think we can do it as part of a different pr and need to support it in our SDK.

arielshaqed commented on pkg/api/controller.go:
```diff
@@ -2937,6 +2937,12 @@ func (c *Controller) handleAPIErrorCallback(ctx context.Context, w http.Response
 	var hookAbortErr *graveler.HookAbortError
 	if errors.As(err, &hookAbortErr) {
+		if errors.Is(hookAbortErr.Unwrap(), kv.ErrSlowDown) {
+			log.Debug("KV Throttling")
```
Sorry, not sure whether this happened.  Could you comment further?  Examples of possible comments (I honestly don't know what's going on, so cannot say):
- "Opened issue #..."
- "I think we don't need a metric, because ..."
- "This is no longer relevant, because...".

arielshaqed replied:
Sorry for very late response, and thanks for reminding me about this! 
Is there a way to see in some metrics how often this happens? For instance, if we end up bumping some API calls metric with a unique status code then that's great. If not, can we count it in some metric? 
Asking because it seems like an important (if extreme) failure mode, and it would be interesting to observe it for users and for us.

Annaseli replied:
@arielshaqed 
Next week, I have a session with Guy about Grafana, so I’ll consult with him on this as well.

Just to confirm to make sure we're aligned on what exactly we want the metric to reflect, the important failure mode you are talking about that we want to monitor is kv.ErrSlowDown, not actions.ErrActionFailed, correct? 

If I understood you correctly, you're interested in tracking how often the “KV Throttling” condition occurs, regardless of whether it originates from a hook abortion or another cause. Since in the current handleAPIErrorCallback function in the controller.go, the case "errors.Is(err, kv.ErrSlowDown)" isn’t specific to hooks.

Annaseli commented:
@N-o-Z 

Currently, when an error comes from actions, we return a 412 (http.StatusPreconditionFailed) status code, even if the error was caused by an internal lakeFS issue. As a result, users don’t receive an accurate status code for internal errors.

The goal of this PR is to distinguish between status codes returned from actions: separating those caused by internal lakeFS errors from those caused by client-side action errors.

I’ve updated the PR title and description to accurately reflect the changes made in this PR, as the previous version was not accurate.

Note that this PR does not specifically handle returning a `503` status code. A 503 status code can be returned during action run if a SlowDown error occurs as any other status code. In such cases, the error is handled by the existing switch statement as all errors are regardless if it came from actions.

Additionally, as I understood from Barak, it was decided before that in the Swagger spec we won't explicitly list all `5xx` errors. Instead, they will fall under the default `500` response. Therefore, Barak doesn't think we need to add `503` explicitly to the response definitions.
